### PR TITLE
[tune] Support `yield` and `return` statements

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

Other libraries use `return` and `yield` statements instead of a custom construct like `tune.report()`. This PR solves this by supporting return statements and generator functions in the function API. This now works:

```

    def trainable(config):
        # config (dict): A dict of hyperparameters.

        for x in range(20):
            intermediate_score = objective(x, config["a"], config["b"])

            yield {"score": intermediate_score}  # This sends the score to Tune.

    analysis = tune.run(
        trainable,
        config={"a": 2, "b": 4}
    )

    print("best config: ", analysis.get_best_config(metric="score", mode="max"))
```

If only a number is yielded (instead of a dict), it is reported to Tune as the anonymous `_metric` metric.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by richardliaw** [comment]
Question - should we also support `tune.report(score)`?

File: python/ray/tune/function_runner.py [hunk 1]
```diff
@@ -495,11 +497,27 @@ class ImplicitFunc(*inherit_from):
...
-                output = train_func(config, reporter)
+                fn = partial(train_func, config, reporter)
+
+            def handle_output(output):
+                if isinstance(output, dict):
+                    reporter(**output)
+                elif isinstance(output, Number):
+                    reporter(_metric=output)
+                else:
+                    pass  # Warn?
```
Comment (richardliaw) [inline python/ray/tune/function_runner.py:512]: I think you should raise here actually

**Review by amogkam — COMMENTED** [review]
Very useful! Looks good to me.

**Comment by krfricke** [comment]
Added support for  anonymous `tune.report(value)`

**Review by richardliaw — APPROVED** [review]

**Comment by richardliaw** [comment]
Looks good. I wonder if in a later PR you can default to `tune.run(metric="_metric")` and raise an error at runtime

**Comment by sumanthratna** [comment]
should we document when users might prefer `return` over `yield`? we can add this to `trainable.rst` and/or to the FAQ.

this might be better suited for another PR

**Comment by richardliaw** [comment]
Tune test failed?

**Comment by krfricke** [comment]
Fixed it, though that didn't really seem related to the PR.

**Comment by richardliaw** [comment]
Hmm that progress output test error looks really suspicious but let's just merge it in and see what happens :)

**Pull request merged**